src/spec.c

```diff
@@ -81,6 +81,55 @@ static void setMemoryInfo(RedisModuleCtx *ctx) {
   RedisModule_FreeServerInfo(ctx, info);
 }
 
+#define MILLISECOND_IN_ONE_SECOND 1000
+#define NANOSECOND_IN_ONE_MILLISECOND 1000000
+
+static inline timespec timespecFromMilliseconds(int64_t totalMilliseconds) {
+  timespec result = {.tv_sec = 0, .tv_nsec = 0};
+  if (totalMilliseconds > 0) {
+    result.tv_sec = totalMilliseconds / MILLISECOND_IN_ONE_SECOND;
+    result.tv_nsec = (totalMilliseconds % MILLISECOND_IN_ONE_SECOND) * NANOSECOND_IN_ONE_MILLISECOND;
+  }
+  return result;
+}
+
+static inline t_expirationTimePoint getDocExpirationTime(RedisModuleCtx* ctx, RedisModuleString *keystr) {
+  timespec zero = {.tv_sec = 0, .tv_nsec = 0};
+  RedisModuleCallReply *rep = RedisModule_Call(ctx, "PEXPIRETIME", "s", keystr);
+  if (rep == NULL || RedisModule_CallReplyType(rep) != REDISMODULE_REPLY_INTEGER) {
+    size_t length = 0;
+    const char* error = rep ? RedisModule_CallReplyStringPtr(rep, &length) : "unknown error";
+    RedisModule_Log(ctx, "warning", "Error calling PEXPIRETIME, type: %d, error(%d): %s", rep ? RedisModule_CallReplyType(rep) : REDISMODULE_REPLY_UNKNOWN, errno, error);
+    return zero;
+  }
+
+  const int64_t totalMilliseconds = RedisModule_CallReplyInteger(rep);
+  t_expirationTimePoint result = timespecFromMilliseconds(totalMilliseconds);
+  RedisModule_FreeCallReply(rep);
+  return result;
+}
+
+static inline FieldExpiration* callHashFieldExpirationTime(RedisModuleCtx* ctx, RedisModuleString *keystr, RedisModuleString** fields) {
+  size_t vectorElementCount = array_len(fields);
+  long long fieldCount = vectorElementCount;
+  RedisModuleCallReply *rep = RedisModule_Call(ctx, "HPEXPIRETIME", "sclv", keystr, "FIELDS", fieldCount, fields, vectorElementCount);
+  if (rep == NULL || RedisModule_CallReplyType(rep) != REDISMODULE_REPLY_ARRAY) {
+    return NULL;
```

Comment:
Not freeing the reply in case it is not NULL (possible?)
When do we get here?

Reply:
I know, seemed odd to me at first.
But it seems like this is a pattern we do across the code.

I guess it can happen if Redis changes the response type for the API.

Reply:
shouldn't happen (in theory we register to version 1 which is the only version at the moment)
Let's at least free the reply if it isn't null